include/gul14/gcd_lcm.h

```diff
@@ -0,0 +1,137 @@
+ *          type, the result is undefined.
+ *
+ * \note
+ * Unlike `std::gcd()` from C++17, the GUL14 version cannot be used with integers of
+ * different signedness. This avoids undefined behavior when mixing unsigned integers with
+ * negative signed values:
+ * \code
+ * // C++17 gcd(): Undefined behavior - the common type is `unsigned int` and -5 has no
+ * // representation in that type
+ * auto bad_result = std::gcd(10u, -5);
+ *
+ * // GUL14 gcd(): Does not compile
+ * auto static_assertion_failure = gul14::gcd(10u, -5);
+ * \endcode
+ *
+ * \since GUL 2.7
+ */
+template <typename IntTypeA, typename IntTypeB>
+constexpr inline auto
+gcd(IntTypeA a, IntTypeB b)
+{
+    static_assert(std::is_integral<IntTypeA>::value and std::is_integral<IntTypeB>::value,
+        "gcd() arguments must be integers");
+
+    static_assert(std::is_signed<IntTypeA>::value == std::is_signed<IntTypeB>::value,
+        "gcd() arguments must have the same signedness");
+
+    using CommonType = std::common_type_t<IntTypeA, IntTypeB>;
+
+    auto c = gul14::abs(static_cast<CommonType>(a));
+    auto d = gul14::abs(static_cast<CommonType>(b));
+
+    while (d != 0)
+    {
+        CommonType tmp = d;
+        d = c % d;
+        c = tmp;
+    }
+
+    return c;
+}
+
+/**
+ * Calculate the least common multiple of two integers.
+ *
+ * If both numbers are zero, the function returns zero. Otherwise, the result is a
+ * positive integer.
+ * \code
+ * int least_common_multiple = lcm(10, 15); // returns 30
+ * \endcode
+ *
+ * \returns the least common multiple of the two numbers, represented as an integer type
+ *          (`std::common_type_t<IntTypeA, IntTypeB>`) that both inputs can implicitly be
+ *          converted to. If either a, b, or the result can not be represented by that
+ *          type, the result is undefined.
+ *
+ * \note
+ * Unlike `std::lcm()` from C++17, the GUL14 version cannot be used with integers of
+ * different signedness. This avoids undefined behavior when mixing unsigned integers with
+ * negative signed values:
+ * \code
+ * // C++17 lcm(): Undefined behavior - the common type is `unsigned int` and -5 has no
+ * // representation in that type
+ * auto bad_result = std::lcm(10u, -5);
+ *
+ * // GUL14 lcm(): Does not compile
+ * auto static_assertion_failure = gul14::lcm(10u, -5);
+ * \endcode
+ *
+ * \since GUL 2.7
+ */
+template <typename IntTypeA, typename IntTypeB>
+constexpr inline auto
+lcm(IntTypeA a, IntTypeB b)
+{
+    static_assert(std::is_integral<IntTypeA>::value and std::is_integral<IntTypeB>::value,
+        "lcm() arguments must be integers");
+
+    static_assert(std::is_signed<IntTypeA>::value == std::is_signed<IntTypeB>::value,
+        "lcm() arguments must have the same signedness");
+
+    using CommonType = std::common_type_t<IntTypeA, IntTypeB>;
+
+    if (a == 0 && b == 0)
+        return CommonType{ 0 };
+
+    return static_cast<CommonType>(gul14::abs(a * b) / gcd(a, b));
```

Comment:
This can overflow for smallish `IntType` and / or moderate values of `a` and `b`. But the docu does not say so.
It might be that it works for `unsigned`, because it will just owerflow the other side again, but for signed integers overlow is UB, iirc.

(Because `a * b`.)

Reply:
Also fails for `unsigned`

```
../tests/test_gcd_lcm.cc:113: FAILED:
  REQUIRE( lcm(60'000u, 100'000u) == 300'000u )
with expansion:
  85251 (0x14d03) == 300000 (0x493e0)
```

Reply:
Very good catch, thanks! I added a commit that uses `intmax_t`/`uintmax_t` for the product and adds documentation on UB should the product not be representable by that type.

Reply:
Found an even better solution, see next thread.